import_obfuscation.go

```diff
@@ -0,0 +1,513 @@
+	importCfg, err := goobj2.ParseImportCfg(importCfgPath)
+	if err != nil {
+		return nil, err
+	}
+	mainPkg, err := goobj2.Parse(objPath, "main", importCfg)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing main objfile: %v", err)
+	}
+	privatePkgs := []pkgInfo{{mainPkg, objPath}}
+
+	// build list of imported packages that are private
+	for pkgPath, info := range importCfg {
+		if isPrivate(pkgPath) {
+			pkg, err := goobj2.Parse(info.Path, pkgPath, importCfg)
+			if err != nil {
+				return nil, fmt.Errorf("error parsing objfile %s at %s: %v", pkgPath, info.Path, err)
+			}
+
+			privatePkgs = append(privatePkgs, pkgInfo{pkg, info.Path})
+		}
+	}
+
+	var sb strings.Builder
+	var buf bytes.Buffer
+	garbledImports := make(map[string]string)
+	for _, p := range privatePkgs {
+		// log.Printf("\n\n++ Obfuscating object file for %s ++\n", p.pkg.ImportPath)
+		for _, am := range p.pkg.ArchiveMembers {
+			// skip objects that are not used by the linker, or that do not contain
+			// any Go symbol info
+			if am.IsCompilerObj() || am.IsDataObj() {
+				continue
+			}
+
+			// add all private import paths to a list to garble
+			var privateImports []string
+			privateImports = append(privateImports, p.pkg.ImportPath)
+			if strings.ContainsRune(p.pkg.ImportPath, '/') {
+				privateImports = append(privateImports, importPathCombos(p.pkg.ImportPath)...)
+			}
+
+			initImport := func(imp string) string {
+				if !isPrivate(imp) {
+					return imp
+				}
+
+				privateImports = append(privateImports, imp)
+				if strings.ContainsRune(imp, '/') {
+					privateImports = append(privateImports, importPathCombos(imp)...)
+				}
+				return hashImport(imp, garbledImports)
+			}
+
+			for i := range am.Imports {
+				am.Imports[i].Pkg = initImport(am.Imports[i].Pkg)
+			}
+			for i := range am.Packages {
+				am.Packages[i] = initImport(am.Packages[i])
+			}
+
+			// move imports that contain another import as a substring to the front,
+			// so that the shorter import will not match first and leak part of an
+			// import path
+			sort.Slice(privateImports, func(i, j int) bool {
+				iSlashes := strings.Count(privateImports[i], "/")
+				jSlashes := strings.Count(privateImports[j], "/")
+				// sort by number of slashes first, then alphabetically
+				if iSlashes == jSlashes {
+					return privateImports[i] > privateImports[j]
+				}
+				return iSlashes > jSlashes
+			})
+			privateImports = dedupImportPaths(privateImports)
+
+			// no private import paths, nothing to garble
+			// log.Printf("== Private imports: %v ==\n", privateImports)
+			if len(privateImports) == 0 {
+				continue
+			}
+
+			// garble all private import paths in all symbol names
+			lists := [][]*goobj2.Sym{am.SymDefs, am.NonPkgSymDefs, am.NonPkgSymRefs}
+			for _, list := range lists {
+				for _, s := range list {
+					// skip debug symbols, they won't affect final binary as we drop debug info
+					if s.Kind >= 8 && s.Kind <= 11 {
```

Comment:
Magic numbers are not good practice. I think `Kind` should be extracted in enum.

Reply:
There is an enum for `Kind`, but it is hidden in an internal Go package. I could expose it from `github.com/Binject/debug/goobj2`

Reply:
for the sake of keeping the internals in a single package, I agree with adding that type to goobj2